File: backend/app/schemas/replay.py
"""
Phase 27 — Historical Replay & Incident Timeline Analysis
Pydantic schemas for the forensic replay engine.
"""

from typing import List, Optional, Dict, Any
from pydantic import BaseModel


class SeverityDistribution(BaseModel):
    CRITICAL: int = 0
    WARNING: int = 0
    INFO: int = 0


class CategoryDistribution(BaseModel):
    POWER: int = 0
    TRAFFIC: int = 0
    WATER: int = 0
    INTERNET: int = 0
    PUBLIC_INFRASTRUCTURE: int = 0


class TimelineBucket(BaseModel):
    """
    A single 30-minute time window in the incident timeline.
    Aggregates all anomaly events that occurred within this window.
    """

    bucket_index: int
    timestamp_start: str
    timestamp_end: str
    label: str  # Human-readable e.g. "May 28, 02:00 – 02:30"
    anomaly_count: int
    critical_count: int
    severity_distribution: SeverityDistribution
    category_distribution: CategoryDistribution
    peak_score: float  # Highest anomaly score in this bucket [0–1]
    health_delta: float  # Change in system health during this window
    event_ids: List[str]  # IDs of incidents in this bucket


class IncidentRecord(BaseModel):
    """
    A single reconstructed incident record for forensic analysis and comparison.
    """

    id: str
    timestamp: str
    metric_name: str
    severity: str
    category: str
    score: float
    description: str
    district: str
    acknowledged: bool
    duration_minutes: int  # Estimated incident duration
    cascaded: bool  # Whether this triggered downstream events
    related_ids: List[str]  # IDs of correlated incidents
    root_cause_hint: Optional[str]  # NLP-derived probable root cause
    resolution_hint: Optional[str]  # Suggested remediation step
    bucket_index: int  # Which timeline bucket this belongs to


class ReplayFrame(BaseModel):
    """
    A single playback frame — snapshot of system state at a specific bucket.
    """

    bucket: TimelineBucket
    active_incidents: List[IncidentRecord]
    cumulative_critical: int
    cumulative_total: int
    system_health: float  # Estimated overall health [0–100] at this moment
    dominant_category: Optional[str]
    alert_level: str  # NOMINAL / ELEVATED / HIGH / CRISIS


class ReplayTimelineResponse(BaseModel):
    """
    Full timeline payload returned for replay initialization.
    """

    buckets: List[TimelineBucket]
    incidents: List[IncidentRecord]
    time_range_hours: int
    bucket_duration_minutes: int
    total_incidents: int
    total_critical: int
    peak_bucket_index: int  # Bucket with most activity
    timeline_start: str
    timeline_end: str


class IncidentComparisonResponse(BaseModel):
    """
    Side-by-side comparison of two incidents for forensic analysis.
    """

    incident_a: IncidentRecord
    incident_b: IncidentRecord
    similarity_score: float  # 0–1, how similar these incidents are
    shared_categories: List[str]
    shared_districts: List[str]
    time_delta_minutes: int  # Time between the two incidents
    likely_correlated: bool
    severity_diff: str  # Which is worse and by how much
    score_diff: float
    combined_risk: str  # Overall risk of both occurring together

File: frontend/src/api/replay.ts
import { getWithRetry } from "./client";

// ==============================================================================
// Response Types
// ==============================================================================

export interface SeverityDistribution {
  CRITICAL: number;
  WARNING: number;
  INFO: number;
}

export interface CategoryDistribution {
  POWER: number;
  TRAFFIC: number;
  WATER: number;
  INTERNET: number;
  PUBLIC_INFRASTRUCTURE: number;
}

export interface TimelineBucket {
  bucket_index: number;
  timestamp_start: string;
  timestamp_end: string;
  label: string;
  anomaly_count: number;
  critical_count: number;
  severity_distribution: SeverityDistribution;
  category_distribution: CategoryDistribution;
  peak_score: number;
  health_delta: number;
  event_ids: string[];
}

export interface IncidentRecord {
  id: string;
  timestamp: string;
  metric_name: string;
  severity: "CRITICAL" | "WARNING" | "INFO";
  category: "POWER" | "TRAFFIC" | "WATER" | "INTERNET" | "PUBLIC_INFRASTRUCTURE";
  score: number;
  description: string;
  district: string;
  acknowledged: boolean;
  duration_minutes: number;
  cascaded: boolean;
  related_ids: string[];
  root_cause_hint: string | null;
  resolution_hint: string | null;
  bucket_index: number;
}

export interface ReplayTimelineResponse {
  buckets: TimelineBucket[];
  incidents: IncidentRecord[];
  time_range_hours: number;
  bucket_duration_minutes: number;
  total_incidents: number;
  total_critical: number;
  peak_bucket_index: number;
  timeline_start: string;
  timeline_end: string;
}

export interface ReplayFrame {
  bucket: TimelineBucket;
  active_incidents: IncidentRecord[];
  cumulative_critical: number;
  cumulative_total: number;
  system_health: number;
  dominant_category: string | null;
  alert_level: "NOMINAL" | "ELEVATED" | "HIGH" | "CRISIS";
}

export interface IncidentComparisonResponse {
  incident_a: IncidentRecord;
  incident_b: IncidentRecord;
  similarity_score: number;
  shared_categories: string[];
  shared_districts: string[];
  time_delta_minutes: number;
  likely_correlated: boolean;
  severity_diff: string;
  score_diff: number;
  combined_risk: "MODERATE" | "HIGH" | "EXTREME";
}

// ==============================================================================
// API Functions
// ==============================================================================

export async function fetchReplayTimeline(hours = 24): Promise<ReplayTimelineResponse> {
  return getWithRetry<ReplayTimelineResponse>(`/api/v1/replay/timeline?hours=${hours}`);
}

export async function fetchReplayFrame(bucketIndex: number, hours = 24): Promise<ReplayFrame> {
  return getWithRetry<ReplayFrame>(`/api/v1/replay/frame/${bucketIndex}?hours=${hours}`);
}

export async function fetchIncidentComparison(
  idA: string,
  idB: string,
  hours = 24
): Promise<IncidentComparisonResponse> {
  return getWithRetry<IncidentComparisonResponse>(
    `/api/v1/replay/compare?id_a=${idA}&id_b=${idB}&hours=${hours}`
  );
}
